packages/ffmpeg/server/FFmpegExporterServer.ts

```diff
@@ -30,20 +54,118 @@ export class FFmpegExporterServer {
     this.command
       .input(this.stream)
       .inputFormat('image2pipe')
       .inputFps(settings.fps);
 
-    // Input audio file
-    if (settings.includeAudio && settings.audio) {
-      this.command
-        .input((settings.audio as string).slice(1))
-        // FIXME Offset only works for negative values.
-        .inputOptions([`-itsoffset ${settings.audioOffset ?? 0}`]);
+    // Input audio
+    const sounds = [...settings.sounds];
+    if (settings.audio && settings.includeAudio) {
+      sounds.push({
+        audio: settings.audio,
+        offset: settings.audioOffset ?? 0,
+      });
+    }
+
+    const filterSpec: FilterSpecification[] = [];
+    const streams: string[] = [];
+
+    const sampleRates: Map<string, number> = new Map();
+    const getRate = async (file: string) => {
+      const rate = sampleRates.get(file);
+      if (rate !== undefined) {
+        return rate;
+      } else {
+        const cmd = `${ffprobePath} -v error -show_entries stream=sample_rate -of default=nw=1:nk=1 ${file.slice(
+          1,
+        )}`;
+        const rate = parseInt((await promisify(exec)(cmd)).stdout);
+        sampleRates.set(file, rate);
+        return rate;
+      }
+    };
+
+    for (let i = 0; i < sounds.length; i++) {
+      const sound = sounds[i];
+      this.command.input(sound.audio.slice(1));
+
+      const filters: AudioVideoFilter[] = [];
+
+      if (sound.start || sound.end !== undefined) {
+        filters.push({
+          filter: 'atrim',
+          options: {start: sound.start, end: sound.end},
+        });
+      }
+
+      if (sound.gain) {
+        filters.push({
+          filter: 'volume',
+          options: {volume: `${sound.gain}dB`},
+        });
+      }
+
+      const playback =
+        Math.pow(2, (sound.detune ?? 0) / 1200) * (sound.playbackRate ?? 1);
+      if (playback !== 1) {
+        const rate = Math.round((await getRate(sound.audio)) * playback);
+        filters.push({
+          filter: 'asetrate',
+          options: {r: rate},
+        });
+      }
+
+      if (sound.offset < 0) {
+        filters.push({
+          filter: 'atrim',
+          options: {start: -sound.offset},
+        });
+      } else if (sound.offset > 0) {
+        const delay = Math.round(sound.offset * 1000);
+        filters.push({
+          filter: 'adelay',
+          options: {delays: delay, all: 1},
+        });
+      }
+
+      if (filters.length > 0) {
+        filterSpec.push({
+          inputs: `${i + 1}:a`,
+          filter: formatFilters(filters),
+          outputs: `a${i + 1}`,
+        });
+        streams.push(`a${i + 1}`);
+      } else {
+        streams.push(`${i + 1}:a`);
+      }
+    }
+
+    if (sounds.length > 0) {
+      this.command.complexFilter([
+        ...filterSpec,
+        {
+          filter: 'amix',
+          // eslint-disable-next-line @typescript-eslint/naming-convention
+          options: {inputs: sounds.length, dropout_transition: 0, normalize: 0},
```

Comment:
Shouldn't `normalize` be `1`? I know next to nothing about ffmpeg though, just asking

Reply:
`normalize: 0` disables the normalization of the `amix` filter, which is intentional
the normalization algorithm that `amix` uses is based on the amount of inputs playing at once, so when an input drops out / ends, it scales the remaining inputs weirdly: https://stackoverflow.com/a/38714779
the best thing to do is normalize the audio afterwards with another (better) normalization filter, or leave it the responsibility of the user to make sure no clipping occurs by adjusting the gain values